### PR TITLE
[MM-24944] Refactor command store to use squirrel

### LINKED ISSUE #14500: Refactor SQL queries in store/sqlstore/command_store.go to use the squirrel builder
Mattermost is replacing the SQL queries of the store layer to use the [squirrel builder](https://github.com/Masterminds/squirrel). This help wanted issue is to change the usages of raw SQL queries in the file `store/sqlstore/command_store.go`.

The expected way to implement this is to search in the `store/sqlstore/command_store.go` file for all the strings containing a SQL query and rewrite them with squirrel. To do so, we must obtain the query builder using the `getQueryBuilder()` method from the store struct, and use it to build the same query that we have in the string.

You can find more information on how to use squirrel in [the official package documentation](https://godoc.org/github.com/Masterminds/squirrel).

Example Pull Request: https://github.com/mattermost/mattermost-server/pull/14393

If you have questions about the ticket or you need any help, feel free to contact `miguel.delacruz` in https://community.mattermost.com/

---
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-24944

### PR DESCRIPTION
**Pull request by dantepippi**
**Summary**

This pull request refactors the command store to use squirrel queries instead of raw SQL queries.

Fixes #14500 
Jira ticket: https://mattermost.atlassian.net/browse/MM-24944

### CODE REVIEW

File: store/sqlstore/command_store.go [hunk 1]
```diff
@@ -83,23 +98,39 @@ func (s SqlCommandStore) GetByTeam(teamId string) ([]*model.Command, *model.AppE
...
 		return nil, model.NewAppError("SqlCommandStore.GetByTrigger", "store.sql_command.get_by_trigger.app_error", nil, "teamId="+teamId+", trigger="+trigger+", err="+err.Error(), http.StatusInternalServerError)
 	}
 
 	return &command, nil
 }
 
 func (s SqlCommandStore) Delete(commandId string, time int64) *model.AppError {
-	_, err := s.GetMaster().Exec("Update Commands SET DeleteAt = :DeleteAt, UpdateAt = :UpdateAt WHERE Id = :Id", map[string]interface{}{"DeleteAt": time, "UpdateAt": time, "Id": commandId})
+	sql, args, err := s.getQueryBuilder().
+		Update("").
```
Comment (ali-farooq0) [inline store/sqlstore/command_store.go:125]: Can this be shortened to:
```
Update("Commands").
SetMap(sq.Eq{"DeleteAt": time, "UpdateAt": time}).
...

```

File: store/sqlstore/command_store.go [hunk 2]
```diff
@@ -108,15 +139,29 @@ func (s SqlCommandStore) Delete(commandId string, time int64) *model.AppError {
 }
 
 func (s SqlCommandStore) PermanentDeleteByTeam(teamId string) *model.AppError {
-	_, err := s.GetMaster().Exec("DELETE FROM Commands WHERE TeamId = :TeamId", map[string]interface{}{"TeamId": teamId})
+	sql, args, err := s.getQueryBuilder().
+		Delete("").
```
Comment (ali-farooq0) [inline store/sqlstore/command_store.go:143]: Also can be shortened to:
```
Delete("Commands").
Where(sq.Eq{"TeamId": teamId}).ToSql()
```

File: store/sqlstore/command_store.go [hunk 3]
```diff
@@ -108,15 +139,29 @@ func (s SqlCommandStore) Delete(commandId string, time int64) *model.AppError {
...
 	if err != nil {
 		return model.NewAppError("SqlCommandStore.DeleteByTeam", "store.sql_command.save.delete_perm.app_error", nil, "id="+teamId+", err="+err.Error(), http.StatusInternalServerError)
 	}
 	return nil
 }
 
 func (s SqlCommandStore) PermanentDeleteByUser(userId string) *model.AppError {
-	_, err := s.GetMaster().Exec("DELETE FROM Commands WHERE CreatorId = :UserId", map[string]interface{}{"UserId": userId})
+	sql, args, err := s.getQueryBuilder().
+		Delete("").
```
Comment (ali-farooq0) [inline store/sqlstore/command_store.go:158]:
```
Delete("Commands").
Where(sq.Eq{"CreatorId": userId}).ToSql()
```

File: store/sqlstore/command_store.go [hunk 4]
```diff
@@ -139,19 +184,20 @@ func (s SqlCommandStore) Update(cmd *model.Command) (*model.Command, *model.AppE
...
+	query := s.getQueryBuilder().
+		Select("COUNT(*)").
+		From("Commands").
+		Where(sq.Eq{"DeleteAt": 0})
 
 	if len(teamId) > 0 {
-		query += " AND TeamId = :TeamId"
+		query = query.Where(sq.Eq{"TeamId": teamId})
+	}
+	queryStr, args, err := query.ToSql()
```
Comment (ali-farooq0) [inline store/sqlstore/command_store.go:195]: nit: missing new line

Comment (ali-farooq0) [reply store/sqlstore/command_store.go:195]: also can we use the same naming convention as above

```
sql, args, err := query.ToSql()
```

File: store/sqlstore/command_store.go [hunk 5]
```diff
@@ -73,8 +82,14 @@ func (s SqlCommandStore) Get(id string) (*model.Command, *model.AppError) {
 
 func (s SqlCommandStore) GetByTeam(teamId string) ([]*model.Command, *model.AppError) {
 	var commands []*model.Command
-
-	if _, err := s.GetReplica().Select(&commands, "SELECT * FROM Commands WHERE TeamId = :TeamId AND DeleteAt = 0", map[string]interface{}{"TeamId": teamId}); err != nil {
+	query, args, err := s.getQueryBuilder().
```
Comment (ali-farooq0) [inline store/sqlstore/command_store.go:85]: minor nit: can we use the same naming convention as your other `ToSql()` calls.
```
sql, args, err := s.getQueryBuilder()....
```

Don't mind `sql` or `query`, as long as they're consistent across all the queries below. :)

File: store/sqlstore/command_store.go [hunk 6]
```diff
@@ -83,23 +98,39 @@ func (s SqlCommandStore) GetByTeam(teamId string) ([]*model.Command, *model.AppE
...
 	var command model.Command
-
-	var query string
+	var triggerStr string
 	if s.DriverName() == "mysql" {
-		query = "SELECT * FROM Commands WHERE TeamId = :TeamId AND `Trigger` = :Trigger AND DeleteAt = 0"
+		triggerStr = "`Trigger`"
 	} else {
-		query = "SELECT * FROM Commands WHERE TeamId = :TeamId AND \"trigger\" = :Trigger AND DeleteAt = 0"
+		triggerStr = "trigger"
```
[the remark this answers is not in the document]
Comment (dantepippi) [reply store/sqlstore/command_store.go:105]: Greetings @ali-farooq0 I'm pretty sure all tests are passing as it was. But I'm ok with the change.
--- PASS: TestCommandStore (0.08s)
    --- PASS: TestCommandStore/MySQL (0.05s)
        --- PASS: TestCommandStore/MySQL/Save (0.00s)
        --- PASS: TestCommandStore/MySQL/Get (0.00s)
        --- PASS: TestCommandStore/MySQL/GetByTeam (0.01s)
        --- PASS: TestCommandStore/MySQL/GetByTrigger (0.01s)
        --- PASS: TestCommandStore/MySQL/Delete (0.01s)
        --- PASS: TestCommandStore/MySQL/DeleteByTeam (0.01s)
        --- PASS: TestCommandStore/MySQL/DeleteByUser (0.01s)
        --- PASS: TestCommandStore/MySQL/Update (0.00s)
        --- PASS: TestCommandStore/MySQL/CommandCount (0.00s)
    --- PASS: TestCommandStore/PostgreSQL (0.03s)
        --- PASS: TestCommandStore/PostgreSQL/Save (0.00s)
        --- PASS: TestCommandStore/PostgreSQL/Get (0.00s)
        --- PASS: TestCommandStore/PostgreSQL/GetByTeam (0.00s)
        --- PASS: TestCommandStore/PostgreSQL/GetByTrigger (0.00s)
        --- PASS: TestCommandStore/PostgreSQL/Delete (0.00s)
        --- PASS: TestCommandStore/PostgreSQL/DeleteByTeam (0.00s)
        --- PASS: TestCommandStore/PostgreSQL/DeleteByUser (0.00s)
        --- PASS: TestCommandStore/PostgreSQL/Update (0.00s)
        --- PASS: TestCommandStore/PostgreSQL/CommandCount (0.00s)
PASS

File: store/sqlstore/command_store.go (hunk 1 above)
Comment (dantepippi) [reply store/sqlstore/command_store.go:125]: Done.

File: store/sqlstore/command_store.go (hunk 2 above)
Comment (dantepippi) [reply store/sqlstore/command_store.go:143]: Done.

File: store/sqlstore/command_store.go (hunk 3 above)
Comment (dantepippi) [reply store/sqlstore/command_store.go:158]: Done.

File: store/sqlstore/command_store.go (hunk 4 above)
Comment (dantepippi) [reply store/sqlstore/command_store.go:195]: Done.

File: store/sqlstore/command_store.go (hunk 5 above)
Comment (dantepippi) [reply store/sqlstore/command_store.go:85]: Done. Going with `sql`.

File: store/sqlstore/command_store.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ali-farooq0) [reply store/sqlstore/command_store.go]: Hey @dantepippi, 
Ahh you're right, looks like even removing `\"trigger\"` from the existing code works. It's possible that this was legacy and has been lingering around for a while. I even tried on my local postgres instance, no issue. But to be on the safe side I say let's stick to `"\"trigger\""`.

File: store/sqlstore/command_store.go [hunk 7]
```diff
@@ -64,7 +66,14 @@ func (s SqlCommandStore) Save(command *model.Command) (*model.Command, *model.Ap
 func (s SqlCommandStore) Get(id string) (*model.Command, *model.AppError) {
 	var command model.Command
 
-	if err := s.GetReplica().SelectOne(&command, "SELECT * FROM Commands WHERE Id = :Id AND DeleteAt = 0", map[string]interface{}{"Id": id}); err != nil {
+	sql, args, err := s.getQueryBuilder().
+		Select("*").
+		From("Commands").
+		Where(sq.Eq{"Id": id, "DeleteAt": 0}).ToSql()
+	if err != nil {
+		return nil, model.NewAppError("SqlCommandStore.Get", "store.sql_command.save.get.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (mgdelacroix) [inline store/sqlstore/command_store.go:74]: For consistency, errors when building queries should use the `store.sql.build_query.app_error` message:

```suggestion
		return nil, model.NewAppError("SqlCommandStore.Get", "store.sql.build_query.app_error", nil, err.Error(), http.StatusInternalServerError)
```

File: store/sqlstore/command_store.go [hunk 8]
```diff
@@ -73,8 +82,14 @@ func (s SqlCommandStore) Get(id string) (*model.Command, *model.AppError) {
...
 func (s SqlCommandStore) GetByTeam(teamId string) ([]*model.Command, *model.AppError) {
 	var commands []*model.Command
-
-	if _, err := s.GetReplica().Select(&commands, "SELECT * FROM Commands WHERE TeamId = :TeamId AND DeleteAt = 0", map[string]interface{}{"TeamId": teamId}); err != nil {
+	sql, args, err := s.getQueryBuilder().
+		Select("*").
+		From("Commands").
+		Where(sq.Eq{"TeamId": teamId, "DeleteAt": 0}).ToSql()
+	if err != nil {
+		return nil, model.NewAppError("SqlCommandStore.GetByTeam", "store.sql_command.save.get_team.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (mgdelacroix) [inline store/sqlstore/command_store.go:90]:
```suggestion
		return nil, model.NewAppError("SqlCommandStore.GetByTeam", "store.sql.build_query.app_error", nil, err.Error(), http.StatusInternalServerError)
```

File: store/sqlstore/command_store.go [hunk 9]
```diff
@@ -83,23 +98,38 @@ func (s SqlCommandStore) GetByTeam(teamId string) ([]*model.Command, *model.AppE
...
+		triggerStr = "\"trigger\""
 	}
 
-	if err := s.GetReplica().SelectOne(&command, query, map[string]interface{}{"TeamId": teamId, "Trigger": trigger}); err != nil {
+	sql, args, err := s.getQueryBuilder().
+		Select("*").
+		From("Commands").
+		Where(sq.Eq{"TeamId": teamId, "DeleteAt": 0, triggerStr: trigger}).ToSql()
+	if err != nil {
+		return nil, model.NewAppError("SqlCommandStore.GetByTrigger", "store.sql_command.get_by_trigger.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (mgdelacroix) [inline store/sqlstore/command_store.go:113]:
```suggestion
		return nil, model.NewAppError("SqlCommandStore.GetByTrigger", "store.sql.build_query.app_error", nil, err.Error(), http.StatusInternalServerError)
```

File: store/sqlstore/command_store.go [hunk 10]
```diff
@@ -83,23 +98,38 @@ func (s SqlCommandStore) GetByTeam(teamId string) ([]*model.Command, *model.AppE
...
 }
 
 func (s SqlCommandStore) Delete(commandId string, time int64) *model.AppError {
-	_, err := s.GetMaster().Exec("Update Commands SET DeleteAt = :DeleteAt, UpdateAt = :UpdateAt WHERE Id = :Id", map[string]interface{}{"DeleteAt": time, "UpdateAt": time, "Id": commandId})
+	sql, args, err := s.getQueryBuilder().
+		Update("Commands").
+		SetMap(sq.Eq{"DeleteAt": time, "UpdateAt": time}).
+		Where(sq.Eq{"Id": commandId}).ToSql()
+	if err != nil {
+		return model.NewAppError("SqlCommandStore.Delete", "store.sql_command.save.delete.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (mgdelacroix) [inline store/sqlstore/command_store.go:129]:
```suggestion
		return model.NewAppError("SqlCommandStore.Delete", "store.sql.build_query.app_error", nil, err.Error(), http.StatusInternalServerError)
```

File: store/sqlstore/command_store.go [hunk 11]
```diff
@@ -108,15 +138,27 @@ func (s SqlCommandStore) Delete(commandId string, time int64) *model.AppError {
 }
 
 func (s SqlCommandStore) PermanentDeleteByTeam(teamId string) *model.AppError {
-	_, err := s.GetMaster().Exec("DELETE FROM Commands WHERE TeamId = :TeamId", map[string]interface{}{"TeamId": teamId})
+	sql, args, err := s.getQueryBuilder().
+		Delete("Commands").
+		Where(sq.Eq{"TeamId": teamId}).ToSql()
+	if err != nil {
+		return model.NewAppError("SqlCommandStore.DeleteByTeam", "store.sql_command.save.delete_perm.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (mgdelacroix) [inline store/sqlstore/command_store.go:145]:
```suggestion
		return model.NewAppError("SqlCommandStore.DeleteByTeam", "store.sql.build_query.app_error", nil, err.Error(), http.StatusInternalServerError)
```

File: store/sqlstore/command_store.go [hunk 12]
```diff
@@ -108,15 +138,27 @@ func (s SqlCommandStore) Delete(commandId string, time int64) *model.AppError {
...
 	return nil
 }
 
 func (s SqlCommandStore) PermanentDeleteByUser(userId string) *model.AppError {
-	_, err := s.GetMaster().Exec("DELETE FROM Commands WHERE CreatorId = :UserId", map[string]interface{}{"UserId": userId})
+	sql, args, err := s.getQueryBuilder().
+		Delete("Commands").
+		Where(sq.Eq{"CreatorId": userId}).ToSql()
+	if err != nil {
+		return model.NewAppError("SqlCommandStore.DeleteByUser", "store.sql_command.save.delete_perm.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (mgdelacroix) [inline store/sqlstore/command_store.go:159]:
```suggestion
		return model.NewAppError("SqlCommandStore.DeleteByUser", "store.sql.build_query.app_error", nil, err.Error(), http.StatusInternalServerError)
```

File: store/sqlstore/command_store.go [hunk 13]
```diff
@@ -139,19 +181,21 @@ func (s SqlCommandStore) Update(cmd *model.Command) (*model.Command, *model.AppE
...
+		Where(sq.Eq{"DeleteAt": 0})
 
 	if len(teamId) > 0 {
-		query += " AND TeamId = :TeamId"
+		query = query.Where(sq.Eq{"TeamId": teamId})
+	}
+
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return 0, model.NewAppError("SqlCommandStore.AnalyticsCommandCount", "store.sql_command.analytics_command_count.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (mgdelacroix) [inline store/sqlstore/command_store.go:195]:
```suggestion
		return 0, model.NewAppError("SqlCommandStore.AnalyticsCommandCount", "store.sql.build_query.app_error", nil, err.Error(), http.StatusInternalServerError)
```

File: store/sqlstore/command_store.go [hunk 14]
```diff
@@ -64,7 +66,14 @@ func (s SqlCommandStore) Save(command *model.Command) (*model.Command, *model.Ap
 func (s SqlCommandStore) Get(id string) (*model.Command, *model.AppError) {
 	var command model.Command
 
-	if err := s.GetReplica().SelectOne(&command, "SELECT * FROM Commands WHERE Id = :Id AND DeleteAt = 0", map[string]interface{}{"Id": id}); err != nil {
+	sql, args, err := s.getQueryBuilder().
+		Select("*").
+		From("Commands").
```
Comment (mgdelacroix) [inline store/sqlstore/command_store.go:71]: Just a nonblocking suggestion, this bit could be abstracted on its own `commandsQuery`, same way as you did in https://github.com/mattermost/mattermost-server/pull/14549

Comment (dantepippi) [reply store/sqlstore/command_store.go]: Done!

File: store/sqlstore/command_store.go [hunk 15]
```diff
@@ -14,11 +14,16 @@ import (
 
 type SqlCommandStore struct {
 	SqlStore
+
+	commandsQuery sq.SelectBuilder
 }
 
 func newSqlCommandStore(sqlStore SqlStore) store.CommandStore {
-	s := &SqlCommandStore{sqlStore}
+	s := &SqlCommandStore{SqlStore: sqlStore}
```
Comment (ali-farooq0) [inline store/sqlstore/command_store.go:23]: minor nit: empty line.

**Comment by teresa-novoa** [comment]
@lindy65 friendly reminder to review this PR, thanks!

**Comment by lindy65** [comment]
Thanks for the reminder @teresa-novoa - I'm just waiting on some testing direction from @mgdelacroix :)

**Pull request merged**